Ships

ccMatrix opened the pull request:
initial work on parsing ship information. let's talk about needed fields and the schema and structure :)

Iliasgx commented:
Some things I noticed that may require some tweaking:

- [ ] The img_url for components is missing
- [ ] For consistency, ship names should be lowercase
- [ ] The additional officer slot should be set to `false` if not applicable
- [ ] The ship_xp is missing for ship levels
- [ ] \u00a0 can be removed from 'type' in costs
- [ ] When scrappable is `true`, required_scrapyard_level should also be present
- [ ] Firing patterns shouldn't contain `null` values

The different weapons are all based on 2 images as far as I know. We don't have to save the same image 30x for all different names.

I thought weapons could have other stats, even with the same name. But checked some and they are the same so I like the change from id to quantity as you presented. 👍 

Not sure why you added build_costs to the ship itself, instead of adding it to a component. Build costs for 'building' the ship are made in T1 since you actually pay for all components to be upgraded to the initial T1.

Ship description could be added, but imo will overthrow a clean design with something people don't really read.

Also presenting both ship cargo info and component cargo info can be confusing for the user. That's why it was added as a single field. Unless you have an idea for this to make use of it.

Lastly, thanks for adding data! 👍

ccMatrix commented:
> * [ ]  \u00a0 can be removed from 'type' in costs

This is actually added by the pre-commit hook on all the files.

> The different weapons are all based on 2 images as far as I know. We don't have to save the same image 30x for all different names.

I went the easy route of just using the component name to save the images to. I can add some more logic there to detect the weapon type and just use something like `component_weapon_energy.png` and `component_weapon_kinetic.png`.

> I thought weapons could have other stats, even with the same name. But checked some and they are the same so I like the change from id to quantity as you presented. 👍

Good to know, I was wondering what the id should be for. So I can just remove id from schema and the files.

> Not sure why you added build_costs to the ship itself, instead of adding it to a component. Build costs for 'building' the ship are made in T1 since you actually pay for all components to be upgraded to the initial T1.

The old data did not have any build and repair info for the T1 components. They were all NIL so I can't populate them. That's why I added the cost of building the initial ship in the root.

> Ship description could be added, but imo will overthrow a clean design with something people don't really read.

I think it doesn't hurt to have that in the data itself. Doesn't mean it must be shown by the bot card.

> Also presenting both ship cargo info and component cargo info can be confusing for the user. That's why it was added as a single field. Unless you have an idea for this to make use of it.

Since the cargo and protected_cargo change after upgrading the component and also after changing tiers I've added both values I found. Don't know how to present them yet but it feels important that a e.g. T3 ship with cargo build has a different cargo capacity than a T4 without cargo build. Again if this is overwhelming info in a bot card it doesn't have to be shown but doesn't hurt to have the info at hand.

WebSpider commented on ships/augur.json:
```diff
@@ -0,0 +1,4749 @@
+  },
+  "ship_ability": {
+    "title": "Obliterate",
+    "description": "As long as the opponent's ship is Burning, if the opponent is hit with the Obliterator weapon, the AUGUR increases all weapon damage by 130.00% (cumulative)."
```
For officers we added abilities like this as "buff". Do we want the same introduction here?

ccMatrix replied:
Should be doable since I already have the code to detect it. As long as the trigger words are the same it should be able to detect that.

WebSpider commented on ships/augur.json:
```diff
@@ -87,6 +87,7 @@
               }
             ]
           },
+          "id": 0,
```
The only usage I can think of for an ID, would be an outside reference. Ie: A system referring to a hostile, or a system referring to an armada. I don't really see what an ID would be doing in a tiered component, tbh.

ccMatrix replied:
I'll remove that field from the files and the schema

Iliasgx replied:
The ID was used since I normally use 'name' as the unique field for every building, research, officer, ship, ... since they are always unique. But components can have multiple weapons with the same name. That's why ID was added. But if we decrease the components and use quantity, the names are again unique.

ccMatrix commented:
It looks like the realta has a bonus that isn't percentage based so I guess we need `percentage: true` in the `ship_ability` object

WebSpider commented on ships/botany_bay.json:
```diff
@@ -0,0 +1,3008 @@
+  },
+  "ship_ability": {
+    "title": "Data Mining",
+    "description": "",
```
Do we have these descriptions somewhere, or do we need to ask the comunity for imput?

WebSpider commented on ships/botany_bay.json:
```diff
@@ -0,0 +1,3008 @@
+        {
+          "name": "Cargo Bay",
+          "additional_info": {
+            "cargo_info": {
```
This is unrelated to the data or the structure, but I'm having trouble interpreting what this says :')

WebSpider commented on ships/vi_dar.json:
```diff
@@ -0,0 +1,3559 @@
+    "scrapable": false
+  },
+  "ship_ability": {
+    "title": "Cybernetic Augmentation The Vi'dar's Damage is increased by 24000% against Borg hostiles.",
```
I think this may need to be split between title and description. The ability is "Cybernetic Augmentation", I think the rest is description

ccMatrix replied:
There was a line break but maybe not in the vidar file. Might need to manually fix that part

WebSpider commented:
Thanks so much for your hard work @ccMatrix . I've made some remarks on seperate lines, but in general a few more points:

- The special material required to upgrade the botany bay and vi'dar seems to be missing. It's the plutonium and charged nanoprobes.
- The meridian seems to be missing. Maybe we dont have complete data on that one?

ccMatrix commented:
I can change the draft to a real PR and we can do a follow-up merge with the manual fixes. Meridian was added after lcars shut down so we don't have any data on it

ccMatrix commented:
I'll check tonight if I have the nanoprobes and plutonium info in there

WebSpider reviewed (approved):
lgtm, with the remark that we are missing some specialty materials for some event ships, but that will be handled in a followup
